tests/0098-consumer-txn.cpp

```diff
@@ -0,0 +1,520 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <iostream>
+#include <cstring>
+#include <cstdlib>
+#include <sys/wait.h>
+#include "testcpp.h"
+#include <assert.h>
+#include <sstream>
+
+
+static void execute_java_produce_cli(std::string &bootstrapServers, std::string &topic, std::string cmd) {
+  std::stringstream ss;
+  ss << "cd java; ./run-class.sh TransactionProducerCli " << bootstrapServers << " " << topic << " " << cmd;
+  system(ss.str().c_str());
+}
+
+static std::vector<RdKafka::Message *> consume_messages(RdKafka::KafkaConsumer *c, std::string topic, int partition) {
+  RdKafka::ErrorCode err;
+
+  /* Assign partitions */
+  std::vector<RdKafka::TopicPartition*> parts;
+  parts.push_back(RdKafka::TopicPartition::create(topic, partition));
+  if ((err = c->assign(parts)))
+    Test::Fail("assign failed: " + RdKafka::err2str(err));
+  RdKafka::TopicPartition::destroy(parts);
+
+  Test::Say("Consuming from topic " + topic + "\n");  
+  std::vector<RdKafka::Message *> result = std::vector<RdKafka::Message *>();
+
+  while (true) {
+    RdKafka::Message *msg = c->consume(tmout_multip(1000));
+    switch (msg->err())
+    {
+      case RdKafka::ERR__TIMED_OUT:
+        continue;
+      case RdKafka::ERR__PARTITION_EOF:
+        break;
+      case RdKafka::ERR_NO_ERROR:
+        result.push_back(msg);
+        continue;
+      default:
+        Test::Fail("Error consuming from topic " + topic + ": " + msg->errstr());
+        break;
+    }
+    break;
+  }
+
+  Test::Say("Read all messages from topic: " + topic + "\n");
+
+  return result;
+}
+
+
+static void delete_messages(std::vector<RdKafka::Message *> &messages) {
+  for (int i=0; i<messages.size(); ++i)
+    delete messages[i];
+}
+
+
+static std::string get_bootstrap_servers() {
+  RdKafka::Conf *conf;
+  std::string bootstrap_servers;
+  Test::conf_init(&conf, NULL, 40);
+  conf->get("bootstrap.servers", bootstrap_servers);
+  delete conf;
+  return bootstrap_servers;
+}
+
+
+static RdKafka::KafkaConsumer *create_consumer(
+    std::string &topic_name,
+    const char *isolation_level) {
+  RdKafka::Conf *conf;
+  std::string errstr;
+
+  Test::conf_init(&conf, NULL, 40);
+  Test::conf_set(conf, "group.id", topic_name);
+  Test::conf_set(conf, "enable.auto.commit", "false");
+  Test::conf_set(conf, "auto.offset.reset", "earliest");
+  Test::conf_set(conf, "enable.partition.eof", "true");
+  Test::conf_set(conf, "isolation.level", isolation_level);
+
+  RdKafka::KafkaConsumer *c = RdKafka::KafkaConsumer::create(conf, errstr);
+  if (!c)
+    Test::Fail("Failed to create KafkaConsumer: " + errstr);
+
+  delete conf;
+
+  return c;
+}
+
+static void test_assert(bool cond, std::string msg) {
```

Comment:
there were complications exposing this to C++ IIRC. I haven't bothered to investigate again since you said you would. pls feel free to suggest required changes.

Reply:
fine for now :+1:  I'll address at a later time